### PR TITLE
Use SString type - PathString for path allocations in binder.

### PR DESCRIPTION
**Pull request by Priya91**
@jkotas @janvorli @JeremyKuhne

### CODE REVIEW

File: src/binder/cdebuglog.cpp [hunk 1]
```diff
@@ -65,8 +65,11 @@ namespace BINDER_SPACE
                 IF_FAIL_GO(HRESULT_FROM_WIN32(ERROR_BUFFER_OVERFLOW));
             }
         
-            IF_FAIL_GO(StringCbCopy(szPath, sizeof(szPath), pszName));
-
+            WCHAR * szPath = szPathString.OpenUnicodeBuffer(MAX_LONGPATH);
```
Comment (jkotas) [inline src/binder/cdebuglog.cpp]: The point of using variable size optimized buffer class is that we should be allocating only what is needed whenever possible. We know exactly on how much we need to allocate here - so we should only allocate as much.

**Comment by prisrirav** [comment]
@jkotas @janvorli @JeremyKuhne : Added commit to replace allocations that are in jeremy's linux longpath change.

**Comment by Priya91** [comment]
@jkotas @janvorli @JeremyKuhne : Added commit to replace allocations that are in jeremy's linux longpath change.

File: src/inc/sstring.h [hunk 2]
```diff
@@ -1006,6 +1006,9 @@ typedef InlineSString<512> StackSString;
 // be needed is small and it's preferable not to take up the stack space.
 typedef InlineSString<32>  SmallStackSString;
 
+// To be used specifically for path strings.
+typedef InlineSString<512> PathString;
```
Comment (jkotas) [inline src/inc/sstring.h]: I would make this 256 (=existing MAX_PATH). Also, I would make the size much smaller in debug builds to make sure that we exercise the buffer allocation path.

```
#ifdef _DEBUG
// Make 
typedef InlineSString<32> PathString; 
#else
typedef InlineSString<256> PathString; 
#endif
```

File: src/binder/utils.cpp [hunk 3]
```diff
@@ -267,14 +267,17 @@ namespace BINDER_SPACE
...
         if (!path.IsEmpty())
         {
-            WCHAR wszCanonicalPath[MAX_LONGPATH];
+            PathString wszCanonicalPathString;
+            WCHAR * pwszCanonicalPath = wszCanonicalPathString.OpenUnicodeBuffer(MAX_LONGPATH);
             PlatformPath(path);
 
             // This is also defined in rotor pal
-            if (PathCanonicalizeW(wszCanonicalPath, path))
+            if (PathCanonicalizeW(pwszCanonicalPath, path))
```
Comment (jkotas) [inline src/binder/utils.cpp]: The contract for `PathCanonicalize` is hard code to `MAX_PATH`. The right fix here is to delete the `BINDER_SPACE::PathCanonizalize` method and its uses - without replacing them with anything. I have reviewed all uses - none of them is required.

File: src/binder/utils.cpp [hunk 4]
```diff
@@ -300,7 +303,9 @@ namespace BINDER_SPACE
         BINDER_LOG_STRING(W("path A"), pathA);
         BINDER_LOG_STRING(W("path B"), pathB);
 
-        WCHAR tempResultPath[MAX_LONGPATH];
+        PathString tempResultPathString;
+        WCHAR * tempResultPath = tempResultPathString.OpenUnicodeBuffer(MAX_LONGPATH);
+
         if (PathCombineW(tempResultPath, pathA, pathB))
```
Comment (jkotas) [inline src/binder/utils.cpp]: The contract for `PathCombine`  is hardcoded to `MAX_PATH`. The right fix here is to implement a simple path combining here, without calling PathCombine:
- Copy `pathA` to `combinedPath`
- Append path separator to `combinedPath` if it does not end with one
- Append `pathB` to `combinedPath`

File: src/classlibnative/bcltype/system.cpp [hunk 5]
```diff
@@ -288,11 +289,13 @@ FCIMPL0(StringObject*, SystemNative::_GetModuleFileName)
 {
     FCALL_CONTRACT;
 
-    WCHAR wszFile[MAX_LONGPATH];
+    PathString wszFilePathString;
+    
     STRINGREF   refRetVal   = NULL;
```
Comment (jkotas) [inline src/classlibnative/bcltype/system.cpp]: There are very subtle rules that need to be following during implementation of FCall. The ones violated here are:
- stack allocated (ie `PathString` objects with constructors has to be inside `HELPER_METHOD_FRAME` scope
- exception has to be thrown inside `HELPER_METHOD_FRAME` scope (`GetUnicode()` can throw)

Comment (jkotas) [inline src/classlibnative/bcltype/system.cpp]: Here is how to fix the implementation:

```
FCIMPL0(StringObject*, SystemNative::_GetModuleFileName)
{
    FCALL_CONTRACT;

    STRINGREF   refRetVal = NULL;

    HELPER_METHOD_FRAME_BEGIN_RET_1(refRetVal);
    if (g_pCachedModuleFileName)
    {
        refRetVal = StringObject::NewString(g_pCachedModuleFileName);
    }
    else
    {
        PathString wszFilePathString;

        WCHAR * wszFile = wszFilePathString.OpenUnicodeBuffer(MAX_LONGPATH);
        DWORD lgth = WszGetModuleFileName(NULL, wszFile, MAX_LONGPATH);
        if (!lgth)
        {
            COMPlusThrowWin32();
        }
        wszFilePathString.CloseBuffer(lgth);

        refRetVal = StringObject::NewString(wszFilePathString.GetUnicode());
    }
    HELPER_METHOD_FRAME_END();

    return (StringObject*)OBJECTREFToObject(refRetVal);
}
FCIMPLEND
```

File: src/classlibnative/bcltype/system.cpp [hunk 6]
```diff
@@ -347,14 +356,16 @@ FCIMPL0(StringObject*, SystemNative::GetRuntimeDirectory)
 {
     FCALL_CONTRACT;
 
-    wchar_t wszFile[MAX_LONGPATH+1];
+    PathString wszFilePathString;
```
Comment (jkotas) [inline src/classlibnative/bcltype/system.cpp]: Could you please move this after the `HELPER_METHOD_FRAME_BEGIN`

File: src/debug/ee/debugger.cpp [hunk 7]
```diff
@@ -15136,7 +15136,8 @@ HRESULT Debugger::InitAppDomainIPC(void)
     } hEnsureCleanup(this);
 
     DWORD dwStrLen = 0;
-    WCHAR szExeName[MAX_LONGPATH];
+    PathString szExeNamePathString;
+    WCHAR * szExeName = szExeNamePathString.OpenUnicodeBuffer(MAX_LONGPATH);
```
Comment (jkotas) [inline src/debug/ee/debugger.cpp]: Note that it would be better to just use regular SString here that always allocates on heap. The PathString stack-allocated buffer will never kick in here because of we always allocate MAX_LONGPATH.

Comment (jkotas) [inline src/debug/ee/debugger.cpp]: There are other places in this delta where the same comment applies

File: src/dlls/mscoree/mscoree.cpp [hunk 8]
```diff
@@ -1137,13 +1141,22 @@ STDAPI LoadLibraryShimInternal(LPCWSTR szDllName, LPCWSTR szVersion, LPVOID pvRe
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: This method is not used in CoreCLR. It would be better to ifdef the whole method with `#ifndef FEATURE_CORECLR` instead of trying to "fix it".

Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: (You can delete the FEATURE_CORECLR specific path in the implementation of the method while you are on it.)

File: src/dlls/mscoree/mscoree.cpp [hunk 9]
```diff
@@ -1196,7 +1210,7 @@ STDAPI LoadLibraryShimInternal(LPCWSTR szDllName, LPCWSTR szVersion, LPVOID pvRe
 #endif // CROSSGEN_COMPILE
 
 static DWORD g_dwSystemDirectory = 0;
-static WCHAR g_pSystemDirectory[MAX_LONGPATH + 1];
+static PathString g_pSystemDirectoryPathString;
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: This code is trying to be lock-free thread safe (see comment `use local buffer for thread safety` below). Assignment of PathString is not thread safe. You should change g_pSystemDirectory to be WCHAR *, allocate buffer of the right size using malloc, copy the result into it, and assign it to g_pSystemDirectory (pointer assignment is atomic).

**Comment by jkotas** [comment]
@Priya91 I went through all changes and left comments.

File: src/debug/ee/debugger.cpp (hunk 7 above)
Comment (Priya91) [inline src/debug/ee/debugger.cpp]: @jkotas The PathString is InlineSString<512>, and 512 is the size of stack allocated buffer in InlineSString, anything greater than 512 is allocated with new Byte[buffersize] which should be on heap. OpenUnicodeBuffer(bufSize) is the maximum size we expect to write to the buffer, and on CloseBuffer(bufSize), we tell exactly the size to truncate the string. Is my understanding wrong, as far as I saw in code, the OpenUnicodeBuffer(size) does not put the whole buffer on stack.

Comment (jkotas) [inline src/debug/ee/debugger.cpp]: The whole buffer is always either on stack or on heap. It is on stack only if requested size <= reserved size - it will never be the case here because of requested size = MAX_LONGPATH and reserved size = 512.

**Comment by Priya91** [comment]
@jkotas I've submitted a new commit with the PR feedback, please review for merge. Thanks!

File: src/dlls/mscoree/mscoree.cpp [hunk 10]
```diff
@@ -792,7 +792,13 @@ BOOL PAL_GetPALDirectory(__out_ecount(cchBuffer) LPWSTR pbuffer,
 
     HRESULT hr = S_OK;
 
-    WCHAR pPath[MAX_LONGPATH];
+    WCHAR * pPath = new (nothrow) WCHAR[MAX_LONGPATH];
+    if (pPath == NULL)
+    {
+        WCHAR pPathStack[MAX_LONGPATH];
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: The error handling here should be:
```
SetLastError(ERROR_NOT_ENOUGH_MEMORY);
return FALSE;
```

Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: `pPathStack` in your original code is guaranteed to be allocated on stack only by the closing `}` of the scope - `pPath` would be a dangling pointer.

Comment (janvorli) [inline src/dlls/mscoree/mscoree.cpp]: I am not sure I understand the purpose of this construct. The pPathStack will be allocated always on the stack, so why do we allocate the pPath at all? And for compilers that would reuse the frame space of the pPathStack after it goes out of scope at the end of the block, this is even dangerous.

Comment (Priya91) [inline src/dlls/mscoree/mscoree.cpp]: Yup, I just realised that. Didn't know how to handle NULL here.

File: src/binder/utils.cpp [hunk 11]
```diff
@@ -267,7 +267,8 @@ namespace BINDER_SPACE
 
         if (!path.IsEmpty())
         {
-            WCHAR wszCanonicalPath[MAX_LONGPATH];
+            // PathCanonicalizeW is hard coded as MAX_PATH on windows.
+            WCHAR wszCanonicalPath[MAX_PATH];
```
Comment (jkotas) [inline src/binder/utils.cpp]: Would it be possible to just delete this whole function?

File: src/dlls/mscoree/mscoree.cpp [hunk 12]
```diff
@@ -1263,17 +1237,27 @@ HRESULT SetInternalSystemDirectory()
         DWORD len;
 
         // use local buffer for thread safety
-        WCHAR wzSystemDirectory[COUNTOF(g_pSystemDirectory)];
+        WCHAR * wzSystemDirectory = new (nothrow) WCHAR[MAX_LONGPATH+1];
+        if (wzSystemDirectory == NULL)
+        {
+            WCHAR wzSystemDirectoryStack[MAX_LONGPATH+1];
+            wzSystemDirectory = wzSystemDirectoryStack;
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: Same comment about error handling...

File: src/dlls/mscoree/mscoree.cpp [hunk 13]
```diff
@@ -1188,15 +1165,12 @@ STDAPI LoadLibraryShimInternal(LPCWSTR szDllName, LPCWSTR szVersion, LPVOID pvRe
...
-
 }
-
-#endif // CROSSGEN_COMPILE
+#endif
+#endif 
 
 static DWORD g_dwSystemDirectory = 0;
-static WCHAR g_pSystemDirectory[MAX_LONGPATH + 1];
+static WCHAR * g_pSystemDirectory = new WCHAR[MAX_LONGPATH+1];
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: I would initialize `g_pSystemDirectory` to `NULL` here...

File: src/dlls/mscoree/mscoree.cpp [hunk 14]
```diff
@@ -1263,17 +1237,27 @@ HRESULT SetInternalSystemDirectory()
...
 
         if(FAILED(hr)) {
-            wzSystemDirectory[0] = W('\0');
+            *wzSystemDirectory = W('\0');
             len = 1;
         }
-
+        
         // publish results idempotently with correct memory ordering
         memcpy(g_pSystemDirectory, wzSystemDirectory, len * sizeof(WCHAR));
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: ... allocate the buffer of just the right size here, fill it in with the path, and then publish it to g_pSystemDirectory. e.g:
```
WCHAR * pSystemDirectory = new (nothrow) WCHAR[len + 1];
... error handling here...
wcscpy_s(pSystemDirectory, len + 1, wzDirectory);
g_pSystemDirectory = pSystemDirectory;
```

File: src/binder/utils.cpp (hunk 11 above)
Comment (Priya91) [inline src/binder/utils.cpp]: I've removed all the uses, so it can be deleted.

File: src/dlls/mscoree/mscoree.cpp (hunk 12 above)
Comment (Priya91) [inline src/dlls/mscoree/mscoree.cpp]: fixed.

File: src/dlls/mscoree/mscoree.cpp (hunk 13 above)
Comment (Priya91) [inline src/dlls/mscoree/mscoree.cpp]: On initial assigment in ```SetInternalSystemDirectory``` if this pointer doesn't point to allocated memory wouldn't memcpy fail?

File: src/dlls/mscoree/mscoree.cpp (hunk 14 above)
Comment (Priya91) [inline src/dlls/mscoree/mscoree.cpp]: @jkotas Sure, why do this instead of memcpy?

Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: Your original code would allocate MAX_PATHLONG buffer for the duration of the program, that will be mostly unused in the typical case (the typical path will be much smaller than MAX_PATHLONG). This will avoid wasting memory - the buffer allocated for the duration of the program will be of just the right size.

Comment (Priya91) [inline src/dlls/mscoree/mscoree.cpp]: Ok sounds good, so it doesn't matter if the ```g_pSystemDirectory``` is null initially. I've submitted another commit with the suggested changes.

**Comment by Priya91** [comment]
@jkotas @janvorli @JeremyKuhne rebased the feedback commits into a single commit. When this passes CI run, is it good to merge?

File: src/binder/utils.cpp [hunk 15]
```diff
@@ -300,17 +269,21 @@ namespace BINDER_SPACE
...
         BINDER_LOG_STRING(W("path B"), pathB);
 
-        WCHAR tempResultPath[MAX_LONGPATH];
-        if (PathCombineW(tempResultPath, pathA, pathB))
-        {
-            combinedPath.Set(tempResultPath);
-            BINDER_LOG_STRING(W("combined path"), tempResultPath);
-        }
-        else
+        PathString tempResultPath;
```
Comment (janvorli) [inline src/binder/utils.cpp]: We could combine it directly into the combinedPath instead of creating the tempResultPath.

**Comment by Priya91** [comment]
@dotnet-bot test this please

**Comment by Priya91** [comment]
```
"d:\j\workspace\dotnet_coreclr_windows_debug_prtest\bin\obj\Windows_NT.x64.Debug\install.vcxproj" (default target) (1) ->
00:39:40        "D:\j\workspace\dotnet_coreclr_windows_debug_prtest\bin\obj\Windows_NT.x64.Debug\ALL_BUILD.vcxproj" (default target) (2) ->
00:39:40        "D:\j\workspace\dotnet_coreclr_windows_debug_prtest\bin\obj\Windows_NT.x64.Debug\src\dlls\mscoree\coreclr\coreclr.vcxproj" (default target) (20) ->
00:39:40        "D:\j\workspace\dotnet_coreclr_windows_debug_prtest\bin\obj\Windows_NT.x64.Debug\src\binder\v3binder\v3binder.vcxproj" (default target) (56) ->
00:39:40          C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v14.0\CodeAnalysis\Microsoft.CodeAnalysis.targets(214,5): error MSB4175: The task factory "CodeTaskFactory" could not be loaded from the assembly "C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Build.Tasks.Core.dll". Could not find file 'C:\Users\dotnet-bot\AppData\Local\Temp\unr0z0yu.dll'. [D:\j\workspace\dotnet_coreclr_windows_debug_prtest\bin\obj\Windows_NT.x64.Debug\src\binder\v3binder\v3binder.vcxproj]
```

@dotnet-bot test this please.

File: src/dlls/mscoree/mscoree.cpp [hunk 16]
```diff
@@ -792,7 +792,12 @@ BOOL PAL_GetPALDirectory(__out_ecount(cchBuffer) LPWSTR pbuffer,
 
     HRESULT hr = S_OK;
 
-    WCHAR pPath[MAX_LONGPATH];
+    WCHAR * pPath = new (nothrow) WCHAR[MAX_LONGPATH];
+    if (pPath == NULL)
+    {
+        return HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: This should be `return FALSE;` - this function does not return `HRESULT`.

File: src/dlls/mscoree/mscoree.cpp [hunk 17]
```diff
@@ -1263,17 +1233,31 @@ HRESULT SetInternalSystemDirectory()
...
-            wzSystemDirectory[0] = W('\0');
+            *wzSystemDirectory = W('\0');
             len = 1;
         }
-
+        
+        WCHAR * pSystemDirectory = new (nothrow) WCHAR[len];
+        if (pSystemDirectory == NULL)
+        {
+            return HRESULT_FROM_WIN32(ERROR_NOT_ENOUGH_MEMORY);
```
Comment (jkotas) [inline src/dlls/mscoree/mscoree.cpp]: `wzSystemDirectory` will leak if we exit here. Could you please change the line that allocates it to:

`
NewArrayHolder<WCHAR>  wzSystemDirectory(new (nothrow) WCHAR[MAX_LONGPATH+1]);
`

and delete the line that deletes it below. NewArrayHolder will take care of deleting it no matter which path is taken to exit the function.

**Comment by jkotas** [comment]
LGTM othewise

**Comment by Priya91** [comment]
@dotnet-bot test this please
